### ISSUE TITLE
[Feature request] Add a global ThreadPoolManager

### ISSUE THREAD

**Issue by jiming** [threadly/threadly#216]
I think all thread pool should be managed centrally. When I code review code for projects, programmers having been creating thread pool randomly at any place. Which make threads out of control.

In case we have a Global ThreadPoolManager, we can ask program must acquired thread pool from the manager with some necessary meta properties. And ThreadPoolManager's status can be watched by JMX.

Of course, ThreadPoolManager have a lot of feature, I can describe some later.

Thanks!

**Comment by jentfoo** [comment]
This is actually a concept I have considered adding to the library for a while now.  Threadly is already well suited for shared pool use with the design being wrapping those pools (though this does not _always_ make sense, sometimes pool segmentation can be good to encourage cpu cache reuse for specific applications).  

I too usually end up implementing something like this for my work projects.  The reason I have not added it to the library itself is I feel there is some dragons with a concept like this.  Specifically once pool reuse becomes a concept within this manager.

Maybe you could give me your thoughts on what an interface for this might look like from your perspective.  If you want to pr an interface that can be a great way to start the back and forth discussion on what a safe general solution might look like.

Thanks!

**Comment by jiming** [comment]
Dear Jentfoo,

Here is some api I considered so far. This is just a pre-alpha version. 

```

import lombok.Data;

import java.util.concurrent.Executor;
import java.util.concurrent.ScheduledExecutorService;

/**
 * Created by liujiming on 2017/8/23.
 */
@Data
public class ThreadPoolManager {

    /**
     * All thread pool are shared, do a customize
     */
    private int sharedPoolInitSize;
    private int sharedPoolMinSize;
    private int sharedPoolMaxSize;


    Executor acquireExecutor(AcquireThreadPool acquire) {
        // create a proxy for executor, so we can do enhancement at will
        // something like share threads between different thread pools
        throw new RuntimeException("Not implemented yet!");
    }

    ScheduledExecutorService acquireScheduleExecutor(AcquireScheduleThreadPool acquire) {
        throw new RuntimeException("Not implemented yet!");
    }

    boolean pauseTask(Runnable r) {
        // pause schedule task r
        throw new RuntimeException("Not implemented yet!");
    }

    boolean resumeTask(Runnable r) {
        // resume schedule task r
        throw new RuntimeException("Not implemented yet!");
    }

    boolean removeTask(Runnable r) {
        // remove schedule task r
        throw new RuntimeException("Not implemented yet!");
    }


    void destroy() {
        // clean up everything
    }
}



/**
 * Estimated duration a task execution took, help better arrange pool resources.
 *
 * This is a hint for ThreadPoolManager to schedule
 *
 * Created by liujiming on 2017/8/23.
 */
public enum TaskLifecycle {
    /**
     * Less than 1 second
     */
    SECOND
    /**
     * Less than 10 seconds
     */
    , LESS_THAN_TEN_SECOND
    /**
     * Less than 1 minute
     */
    , SECONDS
    /**
     * Less than 1 hour
     */
    , MINUTES
    /**
     * More than 1 hour
     */
    , HOURS;



    static TaskLifecycle getProperValue(long milliseconds) {
        if(milliseconds < 1000L) {return SECOND;}
        if(milliseconds < 10000L) {return LESS_THAN_TEN_SECOND;}
        if(milliseconds < 60000L) {return SECONDS;}
        if(milliseconds < 3600000L) {return MINUTES;}
        // if(milliseconds < 3600000L * 24) {return HOURS;}
        return HOURS;
    }

}



/**
 * Created by liujiming on 2017/8/23.
 */
public enum TaskExecuteStrategy {
    ONCE_ABORT      // for on Runnable, only one can be executed at the same time, if another task come, just abandon it
    , ONCE_WAIT     // for on Runnable, only one can be executed at the same time, if another task come, put it into queue
    , CONCURRENT
    ;
}



import lombok.Data;
import org.threadly.concurrent.TaskPriority;

import java.util.function.Function;

/**
 * Created by liujiming on 2017/8/23.
 */
@Data
public class AcquireThreadPool {
    // group that the thread pool belong to, used to category thread pool for better understanding
    String group;
    // name of the thread pool
    String name;

    int maxSize;
    int minSize;

    // rename the current thread to <code>group + "." + name</code> so we can see which task using the thread very easy
    boolean renameThread = true;


    TaskPriority priority;
    TaskExecuteStrategy taskExecuteStrategy = TaskExecuteStrategy.CONCURRENT;

    TaskLifecycle taskLifecycle;

    // call this function before do real task in thread, if return result is FALSE, the task should be canceled
    Function<?, Boolean> preprocess;
    Function<?, Boolean> postprocess;

}




import lombok.Data;

/**
 * Created by liujiming on 2017/8/23.
 */
@Data
public class AcquireScheduleThreadPool extends AcquireThreadPool {

}

```

Have a nice day!
Jiming

**Comment by jentfoo** [comment]
I will definitely be giving this some more thought.  I think one or possibly multiple global pool implementations might be included in `5.4`.  The big problem or "dragon" I worry about is that I feel a pool like this almost has to have some information about the task to do this well.  You are trying to provide that information in expected runtimes, but really that's dancing around the problem (from my perspective, let me know your thoughts).

The real information we need to know about the task is "does it block?"  Meaning if it's computationally bound, then there is no reason to scale the threads beyond the core count of the box (and in fact doing so will only hurt).  If it's blocking, then we do want to scale.  In this case we want to scale based off what the blocking resource is capable of accepting.

In short it seems like to me the ideal design would know if it's blocking or non-blocking, and if blocking then we need to know what it's blocking against, and how much we can hammer that resource in parallel.

This could be done purely through documentation, but I worry about being able to describe it in a clear way to others.  We could potentially extrapolate some of this blocking information based off stack trace analysis (but this may be a significant performance hit).

Let me know if you have any thoughts on the above, otherwise I will continue to think about this myself :)

**Comment by lwahlmeier** [comment]
Here is a quick prototype I came up with.  The main issue is dealing with consumers that might be blocking threads, so anyone asking for a pool needs to tell us how many threads have to be dedicate to it, and how many they want in general.

https://gist.github.com/lwahlmeier/9cdd428501586c9b77bfaf85555304d8

Still probably some concurrency issues with adjusting pool sizes, but trying to do all that w/o blocking.

**Comment by jiming** [comment]
@jentfoo 

Thread blocking is a headache. And I think which is something a thread pool manager shall not concerned, it is the response of the developer.

The key features I think a thread pool manager need to do it as following
* allocate/assign thread to task
    * make three pool? One for HIGH/MEDIUM/LOW priority tasks(A), another only for HIGH priority tasks(B), and last one for HIGH priority tasks on demand(C). 
        * MEDIUM and LOW tasks will wait if the thread pool is exhausted to the max size. 
        * HIGH tasks will try use three pool by sequence. 
            * 1. B
            * 2. A
            * 3. C (C size is 0 - a max size)
    * If all the thread is exhausted or blocked, just let the task wail or abandon

* show very detail information of ThreadPool and threads
    * idle or working,
    * working on which tasks
        * when does it start the current task
        * 

* manageable by jmx

Thanks!

**Comment by jiming** [comment]
@jentfoo 

What I concern most is that the all thread in a system must be manageable and easy to troubleshooting.

We can create hundreds of Excecutor in application from Thread Pool Manager(TPM), but they all share the same thread pool managed by TPM. In that case, we might have declared 300 threads, but actually only 100 thread is created.

Thanks!

**Comment by jiming** [comment]
@lwahlmeier 

Thanks, you are really fast.

What I expect the TPM is that it can handle most of our daily requirement for ThreadPool. So one core part is AcquireThreadPool class in my previous post.

Thank!

**Comment by jentfoo** [comment]
> Thread blocking is a headache. And I think which is something a thread pool manager shall not concerned, it is the response of the developer.

> In that case, we might have declared 300 threads, but actually only 100 thread is created.

In order for us to do your second line safely, we must consider what the developers intent was.  Maybe they needed 300 threads because they all block on external resources.  So I think this is something we need to consider if we are to produce a general tool that is safe for people to depend on.

> The key features I think a thread pool manager need to do it as following

Honestly what you described here was really just a singleton `PriorityScheduler`.  Or am I missing something?

**Comment by jiming** [comment]
@jentfoo 

Maybe I am not cealr about myself previous.

The TPM(ThreadPoolManager) is a global manager of a number of Executor/ExecutorService/ScheduledExecutorService.

When TPM is there we can forbid create Executor/ExecutorService/ScheduledExecutorService by any other methods.

When my code need a Executor/ExecutorService/ScheduledExecutorService, I acquire it from TPM. After that, this code use the Executor/ExecutorService/ScheduledExecutorService instance without caring TPM anymore.

By using standard jdk interface is very helpful for us to integrate with third part lib.

A example for http://projectreactor.io Flux

```

Executor executor = TPM.acquireExecutor(...);

Flux.range(1, 10000) 
    .publishOn(Schedulers.fromExecutor(executor)) 
    .subscribe(result)

```

Without TPM, the code looks like

```

Executor executorCrazy = Executors.newFixedThreadPool(...);

Flux.range(1, 10000) 
    .publishOn(Schedulers.fromExecutor(executorCrazy)) 
    .subscribe(result)

```

The bad thing about later one is that the `executorCrazy` is out of control if a project have hundreds of location need create Executor. Think about it, hundreds of crazy thread pools in project, so horrible.


In case we have TPM who manage all Executor, we can easily check the threads status and locate most online problem related with threads.

Another benefit is that we can allocate less real threads. For example we have 10 excutors, and each of them need 10 thread, then we need create 10 * 10 = 100 threads. With TPM, we can share threads between the 10 excutors, so we might create 50 threads is enough. The poo is share between the 10 excutors.

TPM manage the real thread pool, and excutors looks like thread pool, but themselves have zero thread. When executor need execute a task, it get a thread from TPM thread pool. And release the thread to TPM after its task done.

PriorityScheduller is great, but it is just one Executor, not Executor manager. Perhaps I rename TPM to EM(ExecutorManager) is better?

**Comment by jentfoo** [comment]
> Another benefit is that we can allocate less real threads. 

This is where we need to start considering the behavior and nature of the tasks being submitted.  If we are potentially restricting the threads more then we need to make sure we are doing it in a safe way.

> PriorityScheduller is great, but it is just one Executor

Yea, I did not fully articulate what I meant to you.  What I meant is that you can have a central PriorityScheduler and then adjust it's size as needed.  If you need to return an instance you can either return the same instance, or if you want to prevent complete pool consumption you could wrap it in something like an `ExecutorLimiter` and then return that.  Just because a manager is returning references like they are individual pools, I suspect they would be backed by only one, or a small set of internal pools within the manager.

**Comment by jentfoo** [comment]
@jiming Can you tell me your thoughts on this PR? https://github.com/threadly/threadly/pull/217

In there, the javadocs describe the following public API:

```
public static SchedulerService computationPool();

public static SchedulerService lowPriorityPool(boolean singleThreaded);

public static PrioritySchedulerService singleThreadPool(boolean threadGuaranteed);

public static SchedulerService rangedThreadCountPool(int minThreads, int maxThreads);
```

I think that other types might make sense too (please provide recommendations).  For example in the PR I describe a no-argument pool which polls to inspect the thread state to look for blocking.

Because of how your min threads are "guaranteed", even if other consumers come in and do end up blocking a lot they wont be able to impact the tasks you have requested for execution.

The down side to that is that we must limit every returned pool in order to protect threads which were requested to be.  This results in a bit of overhead, as well as has the potential to waste idle threads.

It's a starting point to hopefully something both you and @lwahlmeier might find useful.

**Comment by jiming** [comment]
@jentfoo 

I have a question. Why not `SchedulerService` implements jdk `ScheduledExecutorService`? IMHO, supply interface from jdk is very import, because some very useful third part lib require them (Executor/ExecutorService/ScheduledExecutorService), or supply a wrapper for them. 

I have some idea about the API you supplied,

- acquired parameters could have more options, please have a look at class `AcquireThreadPool` in my previous comment
- Do not make the method `static`. Perhaps we could define a TPM.getInstance() for global shared instance. And can also supply a new TPM(...) for anther TPM instance. Leave the decision to developers.
- Another thing I didn't see from your API is I might create dozens of Executor by different `AcquireThreadPool` setting. After that I am inject the executor into my business bean. I hope the Executor from TPM is a decorator, so that I can tuning some config at runtime by JMX.


Before I found threadly, I thinks guava's thread utils is best so far. `MoreExecutors`, `Callables.threadRenaming`, and lot's of other highlight, borrow some idea from guava might be a good idea.

Thanks!

**Comment by jentfoo** [comment]
> Why not SchedulerService implements jdk ScheduledExecutorService? IMHO, supply interface from jdk is very import, because some very useful third part lib require them (Executor/ExecutorService/ScheduledExecutorService), or supply a wrapper for them.

Because the reason for threadly pools is to make different compromises from those interfaces.  They require performance compromises that I am not willing to make in threadly.

Most libraries I have encountered do one of two things anyways:
* They also have a custom interface you must implement to (jetty for example, or RX java).
* They only require an `Executor` because their needs are basic.  Which all of threadly libraries do extend this interface.

However we do actually provide the compatibility wrapper's like you described in your question.  So if you must convert to these interfaces, you can: 
http://threadly.github.io/threadly/javadocs/5.3/org/threadly/concurrent/wrapper/compatibility/package-summary.html

They do require a compatibility wrapper specific to the implementation because of the need to be able to get to lower level scheduler functionality in order to implement the JDK's interface with minimal impacts.

**Comment by jentfoo** [comment]
> Do not make the method static. Perhaps we could define a TPM.getInstance() for global shared instance. And can also supply a new TPM(...) for anther TPM instance. Leave the decision to developers.

Wont this just lead to the exact situation you are trying to avoid?  Instead of people creating thread pools all over they start creating TPM's all over?

**Comment by jiming** [comment]
>> Do not make the method static. Perhaps we could define a TPM.getInstance() for global shared instance. And can also supply a new TPM(...) for anther TPM instance. Leave the decision to developers.

> Wont this just lead to the exact situation you are trying to avoid? Instead of people creating thread pools all over they start creating TPM's all over?

Yes, if the developer would like to do it. We can suggest developers to use TPM.getInstance(). But we can also give developer some more options. Perhaps some very complex system do require more than one TPM. 

Anyway, even with multiple TPM, I think most time less than 10 will be enough, better than hundreds of Executors all around.

**Comment by jentfoo** [comment]
Moving this to the 5.5 release version.  I have not forgotten about it.  Instead @lwahlmeier and I have been trying to experiment with some possible performance improvements.  I think we will have some significant changes in 5.4 that we wont want to block behind this work.

My current thoughts here is that we should try to solve one problem first, and then expand later for possibly other use cases.  I am still not convinced that something like this could ever make sense for performance sensitive thread pools.  @jiming let me know if you have had any new thoughts.

**Comment by jiming** [comment]
@jentfoo This is a complicated feature, take your time and refactor step by step please.  

Have a nice day!

**Comment by jentfoo** [comment]
PR #217 was merged to provide this functionality (and will be included in a 5.7 release soon).  Please open new issues if you want changes or improvements to that implementation.

**Comment by jentfoo** [comment]
@jiming I just released `5.16` which does contain a bug fix to this feature if using the `isolatedTask()` functionality.

FWIW, since I am msging you, I was planning to have 5.16 contain some performance improvements for this, but they were not ready by the time that bug was identified.  `5.17` should contain some improvements to this feature you might wan to consume.

Just thought I would let you know since this all came from your idea :)

**Comment by jiming** [comment]
@jentfoo Idea is easy, implementation is difficult. It's all your honor. Thanks again and have a nice day:)
